pygen/pygen_src/riscv_asm_program_gen.py

```diff
@@ -570,10 +572,49 @@ def gen_trap_handler_section(self, hart, mode, cause, tvec,
         if cfg.mtvec_mode == mtvec_mode_t.VECTORED:
             pkg_ins.push_gpr_to_kernel_stack(
                 status, scratch, cfg.mstatus_mprv, cfg.sp, cfg.tp, instr)
+        self.gen_section(pkg_ins.get_label("{}mode_exception_handler".format(mode), hart), instr)
 
     def gen_interrupt_vector_table(self, hart, mode, status, cause, ie,
                                    ip, scratch, instr):
-        pass
+        '''In vector mode, the BASE address is shared between interrupt 0 and exception handling.
+           When vectored interrupts are enabled, interrupt cause 0, which corresponds to user-mode
+           software interrupts, are vectored to the same location as synchronous exceptions. This
+           ambiguity does not arise in practice, since user-mode software interrupts are either
+           disabled or delegated'''
+        instr.extend((".option norvc;", "j {}{}mode_exception_handler".
+                      format(pkg_ins.hart_prefix(hart), mode)))
+        # Redirect the interrupt to the corresponding interrupt handler
+        for i in range(1, rcs.max_interrupt_vector_num):
+            instr.append("j {}{}mode_intr_vector_{}".format(pkg_ins.hart_prefix(hart), mode, i))
+        if not cfg.disable_compressed_instr:
+            instr.append(".option rvc;")
+        for i in range(1, rcs.max_interrupt_vector_num):
+            intr_handler = []
+            pkg_ins.push_gpr_to_kernel_stack(
+                status, scratch, cfg.mstatus_mprv, cfg.sp, cfg.tp, intr_handler)
+            self.gen_signature_handshake(instr=intr_handler,
+                                         signature_type=signature_type_t.CORE_STATUS,
+                                         core_status=core_status_t.HANDLING_IRQ)
+            intr_handler.extend(("csrr x{}, {} # {}".format(
+                                 cfg.gpr[0].value, hex(cause.value), cause.name),
+                                 # Terminate the test if xCause[31] != 0 (indicating exception)
+                                 "srli x{}, x{}, {}".format(
+                                 cfg.gpr[0].value, cfg.gpr[0].value, hex(rcs.XLEN - 1)),
+                                 "beqz x{}, 1f".format(cfg.gpr[0].value)))
+            self.gen_signature_handshake(
+                instr=intr_handler, signature_type=signature_type_t.WRITE_CSR, csr=status)
+            self.gen_signature_handshake(
+                instr=intr_handler, signature_type=signature_type_t.WRITE_CSR, csr=cause)
+            self.gen_signature_handshake(
+                instr=intr_handler, signature_type=signature_type_t.WRITE_CSR, csr=ie)
+            self.gen_signature_handshake(
+                instr=intr_handler, signature_type=signature_type_t.WRITE_CSR, csr=ip)
```

Comment:
Could you please do these four calls to the same function in a for loop for different csrs?

Reply:
fixed!